File: rest-api/src/dda.ts
import type {
  GameAlgoDDAAdjustment,
  GameAlgoDDADecision,
  GameAlgoDDAOptions,
  GameAlgoDDAScriptState,
  GameAlgoExecutionResult,
  GameAlgoStorage,
  JsonValue,
} from "./types.ts";

type BehaviorCounts = Record<string, number>;

type StoredStep = {
  stepId?: string;
  behaviors: BehaviorCounts;
};

type StoredState = {
  schemaVersion: 1;
  current: BehaviorCounts;
  lifetime: BehaviorCounts;
  recentSteps: StoredStep[];
  completedSteps: number;
};

type DDAExecutor = {
  execute(state: JsonValue): Promise<GameAlgoExecutionResult | undefined>;
};

const DEFAULT_WINDOW_SIZE = 10;

export class GameAlgoDDAController {
  private readonly executor: DDAExecutor;
  private readonly storage?: GameAlgoStorage;
  private readonly storageKey: string;
  private readonly recentWindowSize: number;
  private readonly logger?: (message: string) => void;
  private readonly statePromise: Promise<StoredState>;
  private writeTail: Promise<void> = Promise.resolve();

  constructor(
    executor: DDAExecutor,
    storage: GameAlgoStorage | undefined,
    storageKey: string,
    options: GameAlgoDDAOptions = {},
    logger?: (message: string) => void,
  ) {
    this.executor = executor;
    this.storage = storage;
    this.storageKey = options.storageKey ?? storageKey;
    this.recentWindowSize = normalizeWindowSize(options.recentWindowSize);
    this.logger = logger;
    this.statePromise = this.loadState();
  }

  async recordBehavior(type: string, amount = 1): Promise<void> {
    const behaviorType = normalizeBehaviorType(type);
    if (!Number.isFinite(amount) || amount <= 0) throw new Error("DDA behavior amount must be greater than 0");
    await this.mutate((state) => {
      state.current[behaviorType] = (state.current[behaviorType] ?? 0) + amount;
      state.lifetime[behaviorType] = (state.lifetime[behaviorType] ?? 0) + amount;
    });
  }

  async completeStep(stepId?: string): Promise<void> {
    await this.mutate((state) => {
      state.recentSteps.push({
        ...(clean(stepId) ? { stepId: clean(stepId) } : {}),
        behaviors: { ...state.current },
      });
      if (state.recentSteps.length > this.recentWindowSize) {
        state.recentSteps.splice(0, state.recentSteps.length - this.recentWindowSize);
      }
      state.current = {};
      state.completedSteps += 1;
    });
  }

  async reset(scope: "all" | "current" | "recent" = "all"): Promise<void> {
    await this.mutate((state) => {
      if (scope === "all") {
        Object.assign(state, emptyState());
      } else if (scope === "current") {
        state.current = {};
      } else if (scope === "recent") {
        state.recentSteps = [];
      } else {
        throw new Error(`Unsupported DDA reset scope: ${scope}`);
      }
    });
  }

  async snapshot(context: JsonValue = {}): Promise<GameAlgoDDAScriptState> {
    await this.writeTail;
    const state = await this.statePromise;
    const recent = aggregateRecent(state.recentSteps);
    return {
      context,
      behavior: {
        current: { ...state.current },
        recent,
        lifetime: { ...state.lifetime },
        recentSteps: state.recentSteps.map((step) => ({
          ...(step.stepId ? { stepId: step.stepId } : {}),
          behaviors: { ...step.behaviors },
        })),
        completedSteps: state.completedSteps,
        windowSize: this.recentWindowSize,
      },
    };
  }

  async decide(context: JsonValue = {}): Promise<GameAlgoDDADecision> {
    try {
      const execution = await this.executor.execute(await this.snapshot(context));
      if (!execution) return fallbackDecision("executor_not_ready");
      const adjustment = readAdjustment(execution.payload);
      if (!adjustment) return fallbackDecision("invalid_adjustment", execution);
      return {
        adjustment,
        payload: execution.payload,
        diagnostics: execution.diagnostics,
        assignment: execution.assignment,
        isFallback: false,
      };
    } catch (error) {
      this.logger?.(`[GameAlgoSDK] DDA decision failed: ${error instanceof Error ? error.message : String(error)}`);
      return fallbackDecision("execution_failed");
    }
  }

  private async loadState(): Promise<StoredState> {
    try {
      const encoded = await this.storage?.getItem(this.storageKey);
      if (!encoded) return emptyState();
      const state = normalizeStoredState(JSON.parse(encoded));
      if (state.recentSteps.length > this.recentWindowSize) {
        state.recentSteps.splice(0, state.recentSteps.length - this.recentWindowSize);
      }
      return state;
    } catch (error) {
      this.logger?.(`[GameAlgoSDK] DDA state load failed; using empty state: ${error instanceof Error ? error.message : String(error)}`);
      return emptyState();
    }
  }

  private async mutate(update: (state: StoredState) => void): Promise<void> {
    const operation = this.writeTail.then(async () => {
      const state = await this.statePromise;
      update(state);
      await this.storage?.setItem(this.storageKey, JSON.stringify(state));
    });
    this.writeTail = operation.catch(() => undefined);
    return operation;
  }
}

function emptyState(): StoredState {
  return { schemaVersion: 1, current: {}, lifetime: {}, recentSteps: [], completedSteps: 0 };
}

function normalizeStoredState(value: unknown): StoredState {
  if (!value || typeof value !== "object" || Array.isArray(value)) return emptyState();
  const object = value as Record<string, unknown>;
  const recentSteps = Array.isArray(object.recentSteps)
    ? object.recentSteps.flatMap((entry) => {
        if (!entry || typeof entry !== "object" || Array.isArray(entry)) return [];
        const step = entry as Record<string, unknown>;
        return [{
          ...(clean(step.stepId) ? { stepId: clean(step.stepId) } : {}),
          behaviors: normalizeCounts(step.behaviors),
        }];
      })
    : [];
  return {
    schemaVersion: 1,
    current: normalizeCounts(object.current),
    lifetime: normalizeCounts(object.lifetime),
    recentSteps,
    completedSteps: typeof object.completedSteps === "number" && Number.isFinite(object.completedSteps) && object.completedSteps >= 0
      ? Math.floor(object.completedSteps)
      : 0,
  };
}

function normalizeCounts(value: unknown): BehaviorCounts {
  if (!value || typeof value !== "object" || Array.isArray(value)) return {};
  const counts: BehaviorCounts = {};
  for (const [key, amount] of Object.entries(value)) {
    if (key && typeof amount === "number" && Number.isFinite(amount) && amount > 0) counts[key] = amount;
  }
  return counts;
}

function aggregateRecent(steps: StoredStep[]): BehaviorCounts {
  const counts: BehaviorCounts = {};
  for (const step of steps) {
    for (const [type, amount] of Object.entries(step.behaviors)) counts[type] = (counts[type] ?? 0) + amount;
  }
  return counts;
}

function normalizeBehaviorType(value: string): string {
  const type = clean(value);
  if (!type) throw new Error("DDA behavior type is required");
  if (type.length > 128) throw new Error("DDA behavior type must be at most 128 characters");
  return type;
}

function normalizeWindowSize(value?: number): number {
  if (value === undefined) return DEFAULT_WINDOW_SIZE;
  if (!Number.isInteger(value) || value < 1 || value > 100) {
    throw new Error("DDA recentWindowSize must be an integer between 1 and 100");
  }
  return value;
}

function readAdjustment(payload: JsonValue): GameAlgoDDAAdjustment | undefined {
  if (!payload || typeof payload !== "object" || Array.isArray(payload)) return undefined;
  const value = payload.adjustment;
  return value === "increase" || value === "keep" || value === "decrease" ? value : undefined;
}

function fallbackDecision(reason: string, execution?: GameAlgoExecutionResult): GameAlgoDDADecision {
  return {
    adjustment: "keep",
    payload: { adjustment: "keep" },
    diagnostics: { fallback: true, reason },
    ...(execution ? { assignment: execution.assignment } : {}),
    isFallback: true,
  };
}

function clean(value: unknown): string | undefined {
  return typeof value === "string" && value.trim() ? value.trim() : undefined;
}

File: lua/DDA.lua
---@meta
--- Local behavior windows and safe DDA decisions.

local cjson = require("cjson")

local DDA = {}

local function emptyState()
    return {
        schemaVersion = 1,
        current = {},
        lifetime = {},
        recentSteps = {},
        completedSteps = 0,
    }
end

local function copyCounts(source)
    local result = {}
    for key, value in pairs(source or {}) do
        if type(key) == "string" and type(value) == "number" and value > 0 then result[key] = value end
    end
    return result
end

local function normalizeState(value)
    if type(value) ~= "table" then return emptyState() end
    local state = emptyState()
    state.current = copyCounts(value.current)
    state.lifetime = copyCounts(value.lifetime)
    state.completedSteps = math.max(0, math.floor(tonumber(value.completedSteps) or 0))
    if type(value.recentSteps) == "table" then
        for _, step in ipairs(value.recentSteps) do
            if type(step) == "table" then
                table.insert(state.recentSteps, {
                    stepId = step.stepId and tostring(step.stepId) or nil,
                    behaviors = copyCounts(step.behaviors),
                })
            end
        end
    end
    return state
end

local function aggregateRecent(steps)
    local result = {}
    for _, step in ipairs(steps or {}) do
        for behaviorType, amount in pairs(step.behaviors or {}) do
            result[behaviorType] = (result[behaviorType] or 0) + amount
        end
    end
    return result
end

local function fallbackDecision(reason, assignment)
    return {
        adjustment = "keep",
        payload = { adjustment = "keep" },
        diagnostics = { fallback = true, reason = reason },
        assignment = assignment,
        isFallback = true,
    }
end

function DDA.New(options)
    options = options or {}
    local executor = assert(options.executor, "DDA executor is required")
    local storageKey = assert(options.storageKey, "DDA storageKey is required")
    local windowSize = tonumber(options.recentWindowSize) or 10
    if windowSize % 1 ~= 0 or windowSize < 1 or windowSize > 100 then
        error("DDA recentWindowSize must be an integer between 1 and 100")
    end

    local state = emptyState()
    if type(options.storageGet) == "function" then
        local ok, encoded = pcall(options.storageGet, storageKey)
        if ok and encoded and encoded ~= "" then
            local decoded, value = pcall(cjson.decode, encoded)
            if decoded then state = normalizeState(value) end
        end
    end
    while #state.recentSteps > windowSize do table.remove(state.recentSteps, 1) end

    local function persist()
        if type(options.storageSet) ~= "function" then return end
        local encoded, value = pcall(cjson.encode, state)
        if encoded then pcall(options.storageSet, storageKey, value) end
    end

    local controller = {}

    function controller.RecordBehavior(behaviorType, amount)
        behaviorType = tostring(behaviorType or "")
        if behaviorType == "" then error("DDA behavior type is required") end
        if #behaviorType > 128 then error("DDA behavior type must be at most 128 characters") end
        if amount == nil then amount = 1 else amount = tonumber(amount) end
        if not amount or amount ~= amount or amount == math.huge or amount == -math.huge or amount <= 0 then
            error("DDA behavior amount must be a finite number greater than 0")
        end
        state.current[behaviorType] = (state.current[behaviorType] or 0) + amount
        state.lifetime[behaviorType] = (state.lifetime[behaviorType] or 0) + amount
        persist()
    end

    function controller.CompleteStep(stepId)
        table.insert(state.recentSteps, {
            stepId = stepId and tostring(stepId) or nil,
            behaviors = copyCounts(state.current),
        })
        while #state.recentSteps > windowSize do table.remove(state.recentSteps, 1) end
        state.current = {}
        state.completedSteps = state.completedSteps + 1
        persist()
    end

    function controller.Reset(scope)
        scope = scope or "all"
        if scope == "all" then
            state = emptyState()
        elseif scope == "current" then
            state.current = {}
        elseif scope == "recent" then
            state.recentSteps = {}
        else
            error("unsupported DDA reset scope: " .. tostring(scope))
        end
        persist()
    end

    function controller.Snapshot(context)
        local recentSteps = {}
        for _, step in ipairs(state.recentSteps) do
            table.insert(recentSteps, {
                stepId = step.stepId,
                behaviors = copyCounts(step.behaviors),
            })
        end
        return {
            context = type(context) == "table" and context or {},
            behavior = {
                current = copyCounts(state.current),
                recent = aggregateRecent(state.recentSteps),
                lifetime = copyCounts(state.lifetime),
                recentSteps = recentSteps,
                completedSteps = state.completedSteps,
                windowSize = windowSize,
            },
        }
    end

    function controller.Decide(context)
        local ok, result = pcall(executor.Execute, controller.Snapshot(context))
        if not ok or type(result) ~= "table" then return fallbackDecision(ok and "executor_not_ready" or "execution_failed") end
        local payload = result.payload
        local adjustment = type(payload) == "table" and payload.adjustment or nil
        if adjustment ~= "increase" and adjustment ~= "keep" and adjustment ~= "decrease" then
            return fallbackDecision("invalid_adjustment", result.assignment)
        end
        return {
            adjustment = adjustment,
            payload = payload,
            diagnostics = result.diagnostics or {},
            assignment = result.assignment,
            isFallback = false,
        }
    end

    return controller
end

return DDA
